core/src/com/fourinachamber/fortyfive/game/enemy/EnemyAction.kt

```diff
@@ -22,94 +36,132 @@ sealed class EnemyAction(val showProbability: Float, protected val enemy: Enemy)
         predicates.add(predicate)
     }
 
-    fun checkPredicates(controller: GameController): Boolean = predicates.all { it.check(controller) }
+    protected fun checkPredicates(controller: GameController): Boolean = predicates.all { it.check(controller) }
 
-    abstract fun getTimeline(controller: GameController, scale: Double): Timeline
+    abstract fun create(controller: GameController, scale: Double): EnemyAction
 
     open fun applicable(controller: GameController): Boolean = checkPredicates(controller)
 
     class DamagePlayer(
         val damage: IntRange,
         showProbability: Float,
         enemy: Enemy,
-    ) : EnemyAction(showProbability, enemy) {
-
-        override fun getTimeline(controller: GameController, scale: Double): Timeline = Timeline.timeline {
-            include(controller.enemyAttackTimeline(damage.scale(scale * scaleFactor).random()))
+        hasUnlikelyPredicates: Boolean,
+        private val iconHandle: ResourceHandle?,
+    ) : EnemyActionPrototype(showProbability, enemy, hasUnlikelyPredicates) {
+
+        override fun create(controller: GameController, scale: Double): EnemyAction {
+            val damage = damage.scale(scale * scaleFactor).random()
+            return EnemyAction(damage.toString(), iconHandle, this) {
+                include(controller.enemyAttackTimeline(damage))
+            }
         }
-
     }
 
     class DestroyCardsInHand(
         val maxCards: Int,
         showProbability: Float,
         enemy: Enemy,
-    ) : EnemyAction(showProbability, enemy) {
+        hasUnlikelyPredicates: Boolean,
+        private val iconHandle: ResourceHandle?,
+    ) : EnemyActionPrototype(showProbability, enemy, hasUnlikelyPredicates) {
 
         override fun applicable(controller: GameController): Boolean =
             checkPredicates(controller) && controller.cardHand.cards.isNotEmpty()
 
-        override fun getTimeline(controller: GameController, scale: Double): Timeline = Timeline.timeline {
-            var text = ""
-            var amountToDestroy = 0
-            val cardHand = controller.cardHand
-            action {
-                val cardAmount = cardHand.cards.size
-                amountToDestroy = (1..maxCards).random().coerceAtMost(cardAmount - 1)
-                text = TemplateString(
-                    GraphicsConfig.rawTemplateString("destroyCardsInHand"),
-                    mapOf("amount" to amountToDestroy, "s" to if (amountToDestroy == 1) "s" else "")
-                ).string
-            }
-            includeLater(
-                { controller.confirmationPopupTimeline(text) },
-                { true }
-            )
-            action {
+        override fun create(controller: GameController, scale: Double): EnemyAction {
+            val cardAmount = controller.cardHand.cards.size
+            val amountToDestroy = (1..maxCards).random().coerceAtMost(cardAmount - 1)
```

Comment:
why not just irgnore the coerce and put it into the 1..maxCards part? (should be possible i think)

Reply:
I could change it to `(1..maxCards.coerceAtMost(cardAmount - 1)).random()`, but in my opinion that is even harder to read